### PR TITLE
fix(plugin): auth-core now allows authorized users to perform actions…

### PR DESCRIPTION
**Pull request by d-sonuga**
Fix for issue [#7032](https://github.com/amplication/amplication/issues/7032).

### CODE REVIEW

File: plugins/auth-core/src/static/auth/abac.util.ts [hunk 1]
```diff
@@ -9,6 +9,10 @@ export function getInvalidAttributes(
   // eslint-disable-next-line @typescript-eslint/ban-types
   data: Object
 ): string[] {
-  const filteredData = permission.filter(data);
+  // Without this, graphql queries will be denied access
+  // for anything requiring permissions 
+  // regardless of whether or not they have the required permissions.
+  const d = JSON.parse(JSON.stringify(data));
```
Comment (morhag90) [inline plugins/auth-core/src/static/auth/abac.util.ts:15]: @d-sonuga  Please change to :   const filteredData = permission.filter(JSON.parse(JSON.stringify(data))); 
instead of creating the new const d

Comment (overbit) [inline plugins/auth-core/src/static/auth/abac.util.ts:15]: @d-sonuga can you ping point the actual issue? This change might have a performance impact since this function is called on every request and parse(stringify) is quite demanding.

More info:
https://medium.com/ft-product-technology/this-one-line-of-javascript-made-ft-com-10-times-slower-5afb02bfd93f
Potential alternative:  https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-core/src/static/auth/abac.util.ts (hunk 1 above)
Comment (d-sonuga) [reply plugins/auth-core/src/static/auth/abac.util.ts:15]: The root of the issue is the package that's being used to check whether or not a user has permission to carry out an operation on the fields of an entity.

This is the request interceptor from the auth-core plugin that's being used to determine whether or not a user should be allowed to access the fields that are in the request:

```ts
@Injectable()
export class AclValidateRequestInterceptor implements NestInterceptor {
  constructor(
    @InjectRolesBuilder() private readonly rolesBuilder: RolesBuilder,
    private readonly reflector: Reflector
  ) {}

  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    const [permissionsRoles]: any = this.reflector.getAllAndMerge<string[]>(
      "roles",
      [context.getHandler(), context.getClass()]
    );

    const type = context.getType();

    const inputDataToValidate =
      type === "http"
        ? context.switchToHttp().getRequest().body
        : context.getArgByIndex(1).data;

    const permission = this.rolesBuilder.permission({
      role: permissionsRoles.role,
      action: permissionsRoles.action,
      possession: permissionsRoles.possession,
      resource: permissionsRoles.resource,
    });

    const invalidAttributes = abacUtil.getInvalidAttributes(
      permission,
      inputDataToValidate
    );

    if (invalidAttributes.length) {
      throw new ForbiddenException(
        "Insufficient privileges to complete the operation"
      );
    }

    return next.handle();
  }
}
````

To determine the attributes that are invalid, the `abacUtil.getInvalidAttributes` is used:

```ts
import { Permission } from "accesscontrol";

/**
 * @returns attributes not allowed to appear on given data according to given
 * attributeMatchers
 */
export function getInvalidAttributes(
  permission: Permission,
  // eslint-disable-next-line @typescript-eslint/ban-types
  data: Object
): string[] {
  const filteredData = permission.filter(data);
  return Object.keys(data).filter((key) => !(key in filteredData));
}
```

The `getInvalidAttributes` function then uses the `accesscontrol`' package's `Permission.filter` method to determine whether the user has access to the fields accessed in the request body:

The `Permission.filter` function uses a utility function that looks like this:

```ts
/**
     *  Deep clones the source object while filtering its properties by the
     *  given attributes (glob notations). Includes all matched properties and
     *  removes the rest.
     *
     *  @param {Object} object - Object to be filtered.
     *  @param {string[]} attributes - Array of glob notations.
     *
     *  @returns {Object} - Filtered object.
     */
    filter: function (object, attributes) {
        if (!Array.isArray(attributes) || attributes.length === 0) {
            return {};
        }
        var notation = new Notation(object);
        return notation.filter(attributes).value;
    },
```

It uses `Notation.filter` from the `notation` package.

The first few lines of the `Notation.filter` function:

```js
function filter(globNotations) {
            var _this4 = this;

            var original = this.value;
            var copy = _utils2.default.deepCopy(original);

            // ensure array, normalize and sort the globs in logical order. we also
            // concat the array first (to prevent mutating the original) bec. we'll
            // change it's content via `.shift()`
            var globs = _notation2.default.normalize(globNotations).concat();

            // if globs only consist of "*"; set the "copy" as source and return.
            if (_utils2.default.stringOrArrayOf(globs, '*')) {
                this._source = copy;
                return this;
            }
            // if globs is "" or [""] set source to `{}` and return.
            if (arguments.length === 0 || _utils2.default.stringOrArrayOf(globs, '') || _utils2.default.stringOrArrayOf(globs, '!*')) {
                this._source = {};
                return this;
            }
```

It makes a `deepCopy` of  `original` which is the initial permissions object. This works fine on regular objects but on objects with a null prototype, this clone always returns an object with no properties.

This cloning issue is the root of the problem.

In the interceptor, to retrieve the request body:

```js
const inputDataToValidate =
      type === "http"
        ? context.switchToHttp().getRequest().body
        : context.getArgByIndex(1).data;
```

The `context.switchToHttp().getRequest().body` returns a regular object, so the `Permission.filter` works on it correctly but the `context.getArgByIndex(1).data` returns an object with a null prototype. Because the `deepClone` function being used by `Notation.filter` doesn't consider this case, an empty object will always be returned by the `Permission.filter` function meaning that the user doesn't have permission to access any field in the request.

In summary, the root of the problem is the `accesscontrol` package's `Permission.filter`'s incorrect handling of objects with null prototypes.

**Comment by d-sonuga** [comment]
@overbit, I've made the change to use `structuredClone`

**Review by overbit — APPROVED** [review]

**Comment by overbit** [comment]
@d
